src/sage/rings/species.py

```diff
@@ -0,0 +1,1027 @@
+        """
+        Parent.__init__(self, category=InfiniteEnumeratedSets())
+        ElementCache.__init__(self)
+
+    @cached_method
+    def an_element(self):
+        r"""
+        Return an element of ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: C.an_element()
+            [()]
+        """
+        return self._element_constructor_(SymmetricGroup(1))
+
+    def _element_constructor_(self, x):
+        r"""
+        ``x`` is an element of ``self`` or a group `H` such that
+        `H` is directly indecomposable.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: C.clear_cache()
+            sage: C(PermutationGroup([("a", "b", "c")]))
+            [(1,2,3)]
+            sage: C._cache
+            {(3, 3, (3,)): [[(1,2,3)]]}
+            sage: C(PermutationGroup([(1, 3, 5)], domain=[1,3,5]))
+            [(1,2,3)]
+            sage: C(PermutationGroup([[(1,3),(4,7)],[(2,5),(6,8)], [(1,4),(2,5),(3,7)]]))
+            [(5,6)(7,8), (1,2)(3,4), (1,3)(2,4)(5,6)]
+            sage: C._cache
+            {(3, 3, (3,)): [[(1,2,3)]],
+            (8, 8, (2, 2, 4)): [[(5,6)(7,8), (1,2)(3,4), (1,3)(2,4)(5,6)]]}
+        """
+        if parent(x) == self:
+            return x
+        if isinstance(x, PermutationGroup_generic):
+            if len(x.disjoint_direct_product_decomposition()) > 1:
+                raise ValueError(f"{x} is not directly indecomposable")
+            mapping = {v: i for i, v in enumerate(x.domain(), 1)}
+            normalized_gens = [[tuple(mapping[x] for x in cyc)
+                                for cyc in gen.cycle_tuples()]
+                            for gen in x.gens()]
+            P = PermutationGroup(gens=normalized_gens)
+            # Fix for SymmetricGroup(0)
+            if x.degree() == 0:
+                P = SymmetricGroup(0)
+            elm = self.element_class(self, P)
+            return self._cache_get(elm)
+        raise ValueError(f"unable to convert {x} to {self}")
+    
+    def __iter__(self):
+        r"""
+        An iterator over all conjugacy classes of directly indecomposable
+        subgroups.
+
+        TESTS::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: iterC = iter(C)
+            sage: for i in range(5):
+            ....:     print(next(iterC))
+            ....: 
+            []
+            [()]
+            [(1,2)]
+            [(1,2,3)]
+            [(1,2,3), (2,3)]
+        """
+        # Is SymmetricGroup(0) directly indecomposable?
```

Comment:
Yes, according to Chang & Jefferson.  A permutation group $H < \mathfrak S_n$ is called *d.d.p. decomposable* if it has a direct product decomposition $H = H_1 \times \dots \times H_r$ with $r > 1$.  If $H$ consists only of the identity element we have $r = 1$.